util: highlight stack frames

BridgeAR opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
![image](https://user-images.githubusercontent.com/8822573/55379875-ae1c9780-551e-11e9-93dd-10f89ca8d819.png)

Using `util.inspect` on errors is going to highlight userland and
node_module stack frames from now on. This is done by marking Node.js
core frames grey and frames that contain `node_modules` in their path
yellow.

That way it's easy to grasp what frames belong to what code.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented:
I'm guessing those added tests do not run on Windows? I think we'd need to test there as well.

addaleax commented on lib/internal/util/inspect.js:
```diff
@@ -861,10 +862,29 @@ function formatError(err, constructor, tag, ctx) {
+    for (const line of lines) {
+      if (/[/\\]node_modules[/\\]/.test(line)) {
+        newStack += `\n${ctx.stylize(line, 'module')}`;
+      // Node.js core files do not begin with an slash.
```
```suggestion
      // Node.js core files do not begin with a slash.
```

BridgeAR replied:
@bnoordhuis good point. It could also be fooled in case the stack itself is manipulated.

I don't think that there's a perfect solution that is absolutely fool proof but AFAIC that's not a big issue.

addaleax commented:
> That way it's easy to grasp what frames belong to what code.

Looking at your screenshot, it feels like the `node_modules` lines are highlighted over the “regular” ones. I’d either reverse that, so that “own” code is being highlighted, or just mix the two together as one set of unstyled stack frames?

devsnek commented on lib/internal/util/inspect.js:
```diff
@@ -861,10 +862,29 @@ function formatError(err, constructor, tag, ctx) {
+      if (/[/\\]node_modules[/\\]/.test(line)) {
+        newStack += `\n${ctx.stylize(line, 'module')}`;
+      // Node.js core files do not begin with an slash.
+      } else if (/    at .+ \([/\\].+:\d+:\d+\)$/u.test(line)) {
```
do we have a "is internal module" API or something we can use? this seems brittle.

BridgeAR replied:
No, there's none (that I am aware of) and such API could only work for stack traces that have not been tampered with (but that should be fine since it would still be a significant reliability improvement over the current situation).

However, all Node.js versions I know should work with the here used regular expression and I doubt that it's an issue if it rarely fails to properly identify one frame.

BridgeAR commented:
@addaleax 
> I’d either reverse that, so that “own” code is being highlighted, or just mix the two together as one set of unstyled stack frames?

I played around with it and I originally implemented it the other way around. I choose this in the end since yellow is darker than the white frames and the last frame in the stack will likely be a userland frame and that works well with the error message that is not highlighted.
I also tried to highlight only parts of the frame, used underscores and other colors but at least for me highlighting the whole frame in these colors seemed best (so far).
I have no strong opinion on how exactly it should be highlighted and I'll play around with it a bit further and attach a couple of screenshots for comparison.

I am also fine to land this in two PRs: first grey out Node.js core frames and later distinguish `node_modules` from userland code?

addaleax commented:
> I choose this in the end since yellow is darker than the white frames and the last frame in the stack will likely be a userland frame and that works well with the error message that is not highlighted.

That’s really dependent on your terminal and terminal settings, and in particular, we’ve had reports of yellow being hard to read on a bright background.

Maybe it would help if you could explain why it is important to distinguish `node_modules` frames from other userland frames?

BridgeAR commented:
@addaleax 
> That’s really dependent on your terminal and terminal settings

Definitely. Here is an overview for different terminals and how their colors would look like: https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit

Yellow is indeed relatively dark. I originally choose that on purpose as I expect people to mainly focus on their own code and that errors hopefully come mainly from their own code and not from modules that they use.
But I agree that it might be too much in this case. So I changed to be significantly less intrusive by just using an underline for `node_modules` names.

> [...] why [is it] important to distinguish node_modules frames from other userland frames?

I would like to immediately see what node_module was used without having to look more closely at the stack. That's why I differentiated both.

--------

Here a couple of screenshots from alternatives that I checked as comparison. I do not believe that this solution is the best or that there is "a best" solution only that this is an improvement over the current situation. There's still significant room for improvement.

If anyone would rather have a different solution, please let me know! I guess we could just set up a small poll and vote in that case?

![image](https://user-images.githubusercontent.com/8822573/55437646-ab15bb80-559f-11e9-85a5-039c46c4423e.png)
![image](https://user-images.githubusercontent.com/8822573/55437745-f203b100-559f-11e9-9655-84186a691e5b.png)
![image](https://user-images.githubusercontent.com/8822573/55437531-638f2f80-559f-11e9-9cc8-aecae5289617.png)
![image](https://user-images.githubusercontent.com/8822573/55437537-67bb4d00-559f-11e9-9a7d-21f293720389.png)
![image](https://user-images.githubusercontent.com/8822573/55437544-6ee25b00-559f-11e9-889f-76fa88b70d8c.png)
![image](https://user-images.githubusercontent.com/8822573/55437551-76a1ff80-559f-11e9-829e-c54a35f5720a.png)
![image](https://user-images.githubusercontent.com/8822573/55437564-7c97e080-559f-11e9-9f29-f3ce3271e07e.png)
![image](https://user-images.githubusercontent.com/8822573/55437579-8588b200-559f-11e9-99f9-9064ab9443ef.png)

I added another safe guard against detecting non core frames as such by checking if the file is indeed a NativeModule.

@nodejs/util @bnoordhuis @devsnek @addaleax PTAL

BridgeAR commented:
I marked this as semver-minor as it feels like a feature addition.

devsnek commented:
i like the white bold one

BridgeAR commented on doc/api/util.md:
```diff
@@ -631,22 +631,25 @@ via the `util.inspect.styles` and `util.inspect.colors` properties.
 
 Color styling uses ANSI control codes that may not be supported on all
-terminals.
+terminals. To verify color support use [`tty.hasColors()`][].
```
@vsemozhetbyt is our linter working properly? I thought these things would be reported, no?

vsemozhetbyt replied:
IIRC, this case is not caught currently.

BridgeAR commented:
@devsnek 
> i like the white bold one

It's a bit hard to actually identify the differences with it, don't you think so? And  are you also fine with the PR as it is?

BridgeAR commented:
@nodejs/util PTAL. I think this quite a nice feature and it would be great if this could land soonish :)

jasnell commented:
I like the general idea but would make it easier for the user to modify the styles l for each. I know users can already modify the color scheme, but allowing them to select the specific style allows for more variation

BridgeAR commented:
> I [...] would make it easier for the user to modify the styles l for each. [...] allowing them to select the specific style allows for more variation

Can you elaborate? The `color` also stands for e.g., underscores, background color, bold etc.

Do you want the user to also modify the default part (the regular white part)?

BridgeAR commented:
Ping @jasnell. I am not sure what exactly you'd like the user to modify.

Fishrock123 commented:
I like the idea. I think I prefer underlined or bolded `node_modules` names the most.

joyeecheung commented:
This is entirely personal, but from my experience in the stack traces that are really hard to grok, the node_modules ones tend to be the ones you want to ignore. From https://github.com/nodejs/node/pull/27052#issuecomment-479266911 I think they are all highlighting the node modules?

It seems we are having personal preferences regarding how node_modules should be stylized here (and also unsure about the effect in most consoles), maybe one way to move this forward is to only keep the internal frame changes here? I am pretty sure grey dims the frames in most consoles (either dark or bright).

Anyway this is a great idea! Thanks for working on this!

BridgeAR commented:
@joyeecheung 
> [...] the node_modules [stack frames] tend to be the ones you want to ignore [...]. I think they are all highlighting the node modules?

Yes, they currently highlight the `node_modules` by adding underscores under the modules name. We actually want the same (to mainly ignore those). I just could not find a nice middle ground for them right now. With the minimal highlighting it's actually possible to skip those lines while skimming the frames, so it should hopefully mainly result in the same while still being able to easily spot what modules where involved.

> It seems we are having personal preferences regarding how node_modules should be stylized here (and also unsure about the effect in most consoles), maybe one way to move this forward is to only keep the internal frame changes here?

As I outlined above: I do not think that there's an absolute right or wrong but this seems to be an improvement over the current situation and I would like to land it as is. We can always improve this further. The used underscore is very subtle, but if it turns out to be a bad idea, I am all for changing it again.

> Anyway this is a great idea! Thanks for working on this!

Thanks :heart:

BridgeAR commented:
@joyeecheung @jasnell @Fishrock123 @addaleax @devsnek is any of you against landing this as is? That way we could already gather some feedback and improve it further later on.

BridgeAR commented:
Ping @joyeecheung @jasnell @Fishrock123 @addaleax @devsnek @nodejs/util

joyeecheung commented:
I am good with this landing as is, but I don't think I can give it a proper review.

BridgeAR commented:
Thanks for the reviews!

Landed in 693401d0ddd752e5fa47b882e56e252c42c94c0e...ffd2df063c4671e006eba87a0be33357e54964a6 :tada:
